feat: add shells, command, and notebooks launch APIs [DET-4094]

hamidzr opened the pull request:
## Description

- add new apis command, shell, and notebook launch apis
- update config type from byte array to struct (map[string]Value)

<!---
Lead with the intended commit body in this description field. For breaking changes,
please include "BREAKING CHANGE:" at the beginning of your commit body.
At minimum, this section should include a bracketed reference to the Jira ticket,
e.g. "[DET-1234]". When squash-and-merging, copy this directly into the description field.
-->


## Test Plan

- created example code usage through python API bindings (available at the repo) 
- `curl-da.sh http://localhost:8080/api/v1/notebooks -d '{"config": {"resources":{"slots":0}}}'


<!---
Describe the situations in which you've tested your change, and/or a screenshot as appropriate.
Reviewers may ask questions of this test plan to ensure adequate manual coverage of changes.
-->


## Commentary (optional)

<!---
Use this section of your description to add context to the PR. Could be for particularly
tricky bits of code that could use extra scrutiny, historical context useful for reviewers, etc.
You may intentionally leave this section blank and remove the title.
--->


## Checklist

- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.


<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:
- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:
- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:
- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")
-->


[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

hamidzr commented on master/internal/api_command.go:
```diff
@@ -27,3 +40,52 @@ func (a *apiServer) KillCommand(
 	return resp, a.actorRequest(fmt.Sprintf("/commands/%s", req.CommandId), req, &resp)
 }
+
+func (a *apiServer) LaunchCommand(
```
@stoksc any suggestions on how we should get better code reuse between shell and commands? they are so similar even golint is complaining. I explicitly wanted to keep them separate and don't share code between them.

stoksc commented on master/internal/api_shell.go:
```diff
@@ -27,3 +40,52 @@ func (a *apiServer) KillShell(
+	}
+
+	shellID := actorResp.Get().(resourcemanagers.TaskID)
+	shellReq := shellv1.Shell{}
+	actorResp = a.m.system.AskAt(shellsAddr.Child(shellID), &shellReq)
```
Why not
```suggestion
	actorResp = a.m.system.AskAt(shellsAddr.Child(shellID), &shellv1.Shell{})
```

stoksc commented on master/internal/api_command.go:
```diff
@@ -27,3 +40,52 @@ func (a *apiServer) KillCommand(
+		return nil, status.Errorf(codes.Internal, "failed to get the user: %s", err)
+	}
+
+	cmdParams := command.CommandParams{UserFiles: filesToArchive(req.Files)}
+	if req.TemplateName != "" {
+		cmdParams.Template = &req.TemplateName
+	}
+	if req.Config != nil {
+		configBytes, err := protojson.Marshal(req.Config)
+		if err != nil {
+			return nil, err
+		}
+		cmdParams.ConfigBytes = configBytes
+	}
+	if len(req.Data) != 0 {
+		var data map[string]interface{}
+		if err := json.Unmarshal(req.Data, &data); err != nil {
+			return nil, err
+		}
+		cmdParams.Data = data
+	}
```
Re: ci complaining about dupes, for now pull this out into a helper method and pass the individual, shared, inner portions of the LaunchRequests?  Something like command.NewCommandParams() that takes a union of all fields different launch requests provide.

hamidzr replied:
okay I pushed some final changes. I didn't push the helper into the command package to keep proto definitions closer to the API and not into the command package :thinking:

stoksc reviewed (commented):
Looks good so far. Left a comment about how to reduce the duplication. Lmk what you think.

stoksc commented on master/internal/command/tensorboard_manager.go:
```diff
@@ -43,7 +43,7 @@ const (
 // TensorboardRequest describes a request for a new Tensorboard.
 type TensorboardRequest struct {
-	CommandParams
+	*CommandParams
```
Go style nit: I prefer struct embedding only when it embeds behavior not fields, though I'm not sure what the standard is here though. (plus we don't even use anything gained from embedding it).

hamidzr replied:
would the aleternative be to redefine every attribute in commandparams or maybe nest it? the former wouldn't be as DRY

stoksc replied:
It would be like
```suggestion
	CommandParams *CommandParams
```
See https://golang.org/doc/effective_go.html#embedding, in the code it's embedding all the methods and fields of `CommandParams` and they're accessible as if `TensorboardRequest` inherits them.

hamidzr replied:
sg I had updated the shells like this https://github.com/determined-ai/determined/pull/1454/commits/d3f123e5ae237abb8ead30015b87af098ee194d8 I'll make the rest the same.

stoksc commented on master/internal/command/shell_manager.go:
```diff
@@ -52,12 +60,63 @@ func (s *shellManager) Receive(ctx *actor.Context) error {
+	}
+
+	a, _ := ctx.ActorOf(shell.taskID, shell)
+	summaryResponse := ctx.Ask(a, getSummary{})
```
Go style nit: just use `resp` since this is such a short lived var.

stoksc commented on master/internal/command/shell_manager.go:
```diff
@@ -52,12 +60,63 @@ func (s *shellManager) Receive(ctx *actor.Context) error {
+	req ShellLaunchRequest,
+) (*summary, error) {
+	commandReq, err := parseCommandRequest(*req.User, s.db, req.CommandParams,
+		&s.taskSpec.TaskContainerDefaults,
+	)
```
Go style nit: pull this paren up or put all the params on line 82, either way. I wish the formatter handled this, it's really inconsistent all over our code and makes it just a bit harder to skim.

hamidzr replied:
how about 
```
	commandReq, err := parseCommandRequest(
		*req.User, s.db, req.CommandParams,
		&s.taskSpec.TaskContainerDefaults,
	)
```

stoksc commented on master/internal/command/shell_manager.go:
```diff
@@ -52,12 +60,63 @@ func (s *shellManager) Receive(ctx *actor.Context) error {
+			passphrase = &typed
+		}
+	}
+	generatedKeys, err := ssh.GenerateKey(passphrase)
```
Go style nit: just use `key`, since this is such a short lived var. I know this stuff is a stupid comment, but for me it makes a big difference in readability (i'm a bad reader, sorry lol). I won't make this comment anymore :disappear:.

hamidzr replied:
lol no worries, shorter is easie!

stoksc commented on master/internal/command/shell_manager.go:
```diff
@@ -52,12 +60,63 @@ func (s *shellManager) Receive(ctx *actor.Context) error {
+		summary, err := s.processShellLaunchRequest(ctx, msg)
+		if err != nil {
+			ctx.Respond(errors.Wrap(err, "failed to launch shell"))
+		} else {
```
Go style nit: you can drop this else.

stoksc commented on master/internal/command/shell_manager.go:
```diff
@@ -72,40 +131,16 @@ func (s *shellManager) handleAPIRequest(ctx *actor.Context, apiCtx echo.Context)
-
-		shell := s.newShell(req, generatedKeys)
-		if err := check.Validate(shell.config); err != nil {
 			respondBadRequest(ctx, err)
```
For the old API, I think we've lost context on what is a "bad request" error and what is just a 500. I know the old API is deprecated essentially, but so long as the CLI and WebUI use it this could be misleading. Maybe `processShellLaunchRequest` needs an `ok bool` return to show it the request was ok or not?

hamidzr replied:
it I updated it so it returns a status code now.

hamidzr replied:
also good catch!

stoksc commented on master/internal/api_notebook.go:
```diff
@@ -80,3 +85,35 @@ func (a *apiServer) NotebookLogs(
+	}
+
+	notebookID := actorResp.Get().(resourcemanagers.TaskID)
+	actorResp = a.m.system.AskAt(notebooksAddr.Child(notebookID), &notebookv1.Notebook{})
```
Calling `actorResp` something like `notebook` is probably more helpful to the reader. Similar comment elsewhere. Also thought on calling `NotebookLaunchRequest` just `LaunchNotebook`?

hamidzr replied:
re `notebook`

it's a kind of future for notebook. at the time when we assign it we don't know what it is and I  think it'd be if the name reflects that somehow. but as you mention it'd be good if it shows notebook in it somewhere too. it'll just become longer

stoksc replied:
`notebookFut` is ok with me.

stoksc replied:
(but it should be apparent its a future from the type, and go style usually is don't put types in names if you can help it)

hamidzr replied:
hmm some indication of the type in the name can help readability without you having the reader consult the typessytem. it's subjective though

stoksc replied:
yah, i dont mind it, i think it's just one of those go recommended things.

hamidzr replied:
https://github.com/determined-ai/determined/pull/1454/commits/d8569d4a27e87fcc8e6e76e4dfdf0495f0fd6260

stoksc reviewed (commented):
I left comments all over the 4, commands, tensorboards, shells and notebooks, but they mostly apply to all. The only _problem_ (not just a style nit) is that I think the old API will respond with a 400 in a lot of cases it used to not. Lmk what you think.

hamidzr commented:
I addressed the comments and updated the var names for api_shell and shell_managers let me know if those are the only variables you'd like to change in those two files and then I'll go ahead apply the same to 3 other sets : )

stoksc commented:
@hamidzr left comments back for the few questions. Sorry for the nits.

stoksc commented on master/internal/command/shell_manager.go:
```diff
@@ -95,24 +96,24 @@ func (s *shellManager) processShellLaunchRequest(
 	a, _ := ctx.ActorOf(shell.taskID, shell)
-	summaryResponse := ctx.Ask(a, getSummary{})
-	if err = summaryResponse.Error(); err != nil {
+	resp := ctx.Ask(a, getSummary{})
```
I guess I'm an idiot because I said `resp` here and `actorResp` -> `notebookFut` above. Either is fine 🤷

hamidzr replied:
`resp` _is_ shorter :)

stoksc reviewed (approved):
LGTM 👍

stoksc commented:
@hamidzr looks like something broke with TB, but it looks good to me once that's fixed.
